fe/fe-core/src/main/java/org/apache/doris/datasource/hive/HMSTransaction.java

```diff
@@ -0,0 +1,1299 @@
+import org.apache.doris.transaction.Transaction;
+
+import com.google.common.base.Joiner;
+import com.google.common.base.MoreObjects;
+import com.google.common.base.Preconditions;
+import com.google.common.base.Verify;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.Iterables;
+import com.google.common.collect.Lists;
+import com.google.common.collect.Maps;
+import io.airlift.concurrent.MoreFutures;
+import org.apache.hadoop.fs.Path;
+import org.apache.hadoop.hive.metastore.api.Partition;
+import org.apache.hadoop.hive.metastore.api.StorageDescriptor;
+import org.apache.hadoop.hive.metastore.api.Table;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.Queue;
+import java.util.StringJoiner;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentLinkedQueue;
+import java.util.concurrent.Executor;
+import java.util.concurrent.Executors;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.stream.Collectors;
+
+public class HMSTransaction implements Transaction {
+    private static final Logger LOG = LogManager.getLogger(HMSTransaction.class);
+    private final HiveMetadataOps hiveOps;
+    private final RemoteFileSystem fs;
+    private final String dbName;
+    private final String tbName;
+
+    private final Map<DatabaseTableName, Action<TableAndMore>> tableActions = new HashMap<>();
+    private final Map<DatabaseTableName, Map<List<String>, Action<PartitionAndMore>>>
+            partitionActions = new HashMap<>();
+
+    private HmsCommitter hmsCommitter;
+
+    public HMSTransaction(HiveMetadataOps hiveOps, String dbName, String tbName) {
+        this.hiveOps = hiveOps;
+        this.fs = hiveOps.getFs();
+        this.dbName = dbName;
+        this.tbName = tbName;
+    }
+
+    @Override
+    public void commit() {
+        doCommit();
+    }
+
+    public String getDbName() {
+        return dbName;
+    }
+
+    public String getTbName() {
+        return tbName;
+    }
+
+    public List<THivePartitionUpdate> mergePartitions(List<THivePartitionUpdate> hivePUs) {
+        Map<String, THivePartitionUpdate> mm = new HashMap<>();
+        for (THivePartitionUpdate pu : hivePUs) {
+            if (mm.containsKey(pu.getName())) {
+                THivePartitionUpdate old = mm.get(pu.getName());
+                old.setFileSize(old.getFileSize() + pu.getFileSize());
+                old.setRowCount(old.getRowCount() + pu.getRowCount());
+                old.getFileNames().addAll(pu.getFileNames());
+            } else {
+                mm.put(pu.getName(), pu);
+            }
+        }
+        return new ArrayList<>(mm.values());
+    }
+
+    @Override
+    public void rollback() {
+        hmsCommitter.rollback();
```

Comment:
The `hmsCommitter` is init when calling `doCommit()`.
So it may not be init here.